### PR TITLE
Allow add new server names and IPs to apiserver x509 certificate

### LINKED ISSUE #1974: Add new IP to server certificate
**FEATURE REQUEST**

Control `minikube` on remote host through SSH local port forwarding.

**Environment**:
 **Minikube version** (use `minikube version`): v0.22.1
- **VM Driver** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep DriverName`): kvm
- **ISO version** (e.g. `cat ~/.minikube/machines/minikube/config.json | grep -i ISO` or `minikube ssh cat /etc/VERSION`): testing

**What happened**:

I setup `kubectl` on my host to connect to `127.0.0.5`, set up ssh with port forwarding, but connection fails with:
```
Kubernetes master is running at https://127.0.0.5:8443
I0914 19:57:19.537863   11391 cached_discovery.go:126] skipped caching discovery info due to Get https://127.0.0.5:8443/api: x509: certificate is valid for 192.168.42.242, 10.0.0.1, not 127.0.0.5
```

Where the server certificate is located and how to add `127.0.0.5` to it?

### LINKED ISSUE #2361: --apiserver-name should change apiserver name, not CA name
**BUG REPORT**

Please provide the following details:

**Environment**:

**OS**: Ubuntu 16.04.3 LTS

**VM driver**: virtualbox

**ISO version**: minikube-v0.23.6.iso, also tried minikube-v0.24.0.iso

**What happened**:

Command line parameter `--apiserver-name` is advertised as useful "if you want to make the apiserver available from outside the machine". For that to work, you would expect it to change the subject name of `apiserver.crt`. But in practice it changes the issuer/CA name which is not very useful for making the apiserver accessible outside the machine.

Also I can't think of a common scenario where changing the CA name would be desired.

```
$ minikube start --apiserver-name=minikube.example.com
...

$ openssl x509 -noout -issuer -subject -in ~/.minikube/ca.crt
issuer= /CN=minikube.example.com
subject= /CN=minikube.example.com

$ openssl x509 -noout -issuer -subject -in ~/.minikube/apiserver.crt
issuer= /CN=minikube.example.com
subject= /O=system:masters/CN=minikube

$ openssl x509 -noout -issuer -subject -in ~/.minikube/client.crt
issuer= /CN=minikube.example.com
subject= /O=system:masters/CN=minikube-user
```

**What you expected to happen**:

```
$ openssl x509 -noout -issuer -subject -in ~/.minikube/ca.crt
issuer= /CN=minikubeCA
subject= /CN=minikubeCA

$ openssl x509 -noout -issuer -subject -in ~/.minikube/apiserver.crt
issuer= /CN=minikubeCA
subject= /O=system:masters/CN=minikube.example.com

$ openssl x509 -noout -issuer -subject -in ~/.minikube/client.crt
issuer= /CN=minikubeCA
subject= /O=system:masters/CN=minikube-user
```
**How to reproduce it** (as minimally and precisely as possible):

```
$ minikube start --apiserver-name=minikube.example.com
$ IPAddress=$(cat ~/.minikube/machines/minikube/config.json | grep '"IPAddress":' | grep -o '[0-9\.]*')
$ curl https://minikube.example.com:8443 --cacert ~/.minikube/ca.crt --resolve minikube.example.com:8443:$IPAddress
curl: (51) SSL: certificate subject name (minikube) does not match target host name 'minikube.example.com'
```

**Output of `minikube logs` (if applicable)**:

**Anything else do we need to know**:

### PR DESCRIPTION
**Pull request by lichen2013**
Fixes  #2361
Fixes  #1974

### CODE REVIEW

**Comment by dlorenc** [comment]
@minikube-bot ok to test

**Comment by lichen2013** [comment]
I think these failed tests are not really caused by my code.

File: pkg/minikube/bootstrapper/certs.go [hunk 1]
```diff
@@ -144,9 +152,9 @@ func generateCerts(k8s KubernetesConfig) error {
 		{ // apiserver serving cert
 			certPath:       filepath.Join(localPath, "apiserver.crt"),
 			keyPath:        filepath.Join(localPath, "apiserver.key"),
-			subject:        "minikube",
-			ips:            []net.IP{net.ParseIP(k8s.NodeIP), serviceIP, net.ParseIP("10.0.0.1")},
-			alternateNames: util.GetAlternateDNS(k8s.DNSDomain),
+			subject:        k8s.APIServerName,
```
Comment (dlorenc) [inline pkg/minikube/bootstrapper/certs.go]: Can you explain why the "subject" needs to match? Can it stay as "minikube"?

File: cmd/minikube/cmd/start.go [hunk 2]
```diff
@@ -378,6 +382,8 @@ func init() {
 	startCmd.Flags().StringArrayVar(&dockerEnv, "docker-env", nil, "Environment variables to pass to the Docker daemon. (format: key=value)")
 	startCmd.Flags().StringArrayVar(&dockerOpt, "docker-opt", nil, "Specify arbitrary flags to pass to the Docker daemon. (format: key=value)")
 	startCmd.Flags().String(apiServerName, constants.APIServerName, "The apiserver name which is used in the generated certificate for localkube/kubernetes.  This can be used if you want to make the apiserver available from outside the machine")
+	startCmd.Flags().StringArrayVar(&apiServerNames, "apiserver-names", nil, "A set of apiserver names which are used in the generated certificate for localkube/kubernetes.  This can be used if you want to make the apiserver available from outside the machine")
+	startCmd.Flags().StringArrayVar(&apiServerIPs, "apiserver-ips", nil, "A set of apiserver IP Addresses which are used in the generated certificate for localkube/kubernetes.  This can be used if you want to make the apiserver available from outside the machine")
```
Comment (dlorenc) [inline cmd/minikube/cmd/start.go]: Could we use IPSliceVar instead of StringArrayVar?
https://godoc.org/github.com/spf13/pflag#IPSliceVar

Comment (lichen2013) [reply cmd/minikube/cmd/start.go]: sure, will update.

File: pkg/minikube/bootstrapper/certs.go (hunk 1 above)
Comment (lichen2013) [reply pkg/minikube/bootstrapper/certs.go]: The 'subject' change is according to the original issue description. 

But, the goal for this change is to make apiserver accessible outside the machine, and obviously, due to the deprecated usage of Subject CN matching and only check Subject Alternative Names, this is not necessary any more. Will change it back to 'minikube'.

**Comment by haizaar** [comment]
Hi guys,

It would really be helpful to have this working in upstream. Can anyone please help with moving this forward?

**Comment by haizaar** [comment]
Thank you!
Do you know when we can expect new release of minikube?

**Comment by dlorenc** [comment]
We should be cutting a release early next week which will get this out. Thanks!

**Comment by abitrolly** [comment]
@lichen2013 👍 

Is it possible to join params into one `--api-cert-hosts`?

**Pull request merged**